go/pkg/services/log_collector/server.go

```diff
@@ -16,36 +16,507 @@ package log_collector
+	for runId, logItem := range state.InProgress {
+		lcs.Logger.DebugWithCtx(ctx, "Starting log collection for log item", "runId", runId)
+		if _, err := lcs.StartLog(ctx, &log_collector.StartLogRequest{
+			RunId:    runId,
+			Selector: logItem.LabelSelector,
+		}); err != nil {
+			lcs.Logger.WarnWithCtx(ctx, "Failed to start log collection for log item", "runId", runId)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (lcs *LogCollectorServer) RegisterRoutes(ctx context.Context) {
+	lcs.AbstractMlrunGRPCServer.RegisterRoutes(ctx)
+	log_collector.RegisterLogCollectorServer(lcs.Server, lcs)
+}
+
+// StartLog writes the log item info to the state file, gets the pod using the label selector,
+// triggers `monitorPod` and `streamLogs` goroutines.
+func (lcs *LogCollectorServer) StartLog(ctx context.Context, request *log_collector.StartLogRequest) (*log_collector.StartLogResponse, error) {
+
+	lcs.Logger.DebugWithCtx(ctx, "Received Start Log request", "request", request)
+
+	// write log item in progress to file
+	if err := lcs.addItemToInProgress(ctx, request.RunId, request.Selector); err != nil {
+		err := errors.Wrapf(err, "Failed to add run id %s to state file", request.RunId)
+		return &log_collector.StartLogResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	var pods *v1.PodList
+	var err error
+
+	// list pods using label selector until a pod is found
+	for len(pods.Items) == 0 {
+		// TODO: add timeout to this loop
+
+		pods, err = lcs.kubeClientSet.CoreV1().Pods(lcs.namespace).List(ctx, metav1.ListOptions{
+			LabelSelector: request.Selector,
+		})
+		if err != nil {
+			err := errors.Wrapf(err, "Failed to list pods for run id %s", request.RunId)
+			return &log_collector.StartLogResponse{
+				Success: false,
+				Error:   err.Error(),
+			}, err
+		}
+	}
+
+	// found a pod. for now, we only assume each run has a single pod.
+	pod := pods.Items[0]
+
+	// TODO: create a child context before calling goroutines, so it won't be canceled
+
+	// stream logs to file
+	go lcs.streamPodLogs(ctx, request.RunId, pod.Name)
+
 	return &log_collector.StartLogResponse{
 		Success: true,
-		Error:   "",
 	}, nil
 }
 
-func NewLogCollectorServer(logger *loggerus.Loggerus) (*logCollectorServer, error) {
-	abstractServer, err := framework.NewAbstractMlrunGRPCServer(logger, nil)
+// GetLog returns the log file contents of length size from an offset, for a given run id
+func (lcs *LogCollectorServer) GetLog(ctx context.Context, request *log_collector.GetLogRequest) (*log_collector.GetLogResponse, error) {
+	lcs.Logger.DebugWithCtx(ctx, "Received Get Log request", "request", request)
+
+	// get log file path
+	filePath, err := lcs.getLogFilePath(request.RunId)
 	if err != nil {
-		return nil, errors.Wrap(err, "Failed to create abstract server")
+		err := errors.Wrapf(err, "Failed to get log file path for run id %s", request.RunId)
+		return &log_collector.GetLogResponse{
+			Success: false,
+			Error:   err.Error(),
```

Comment:
why not retuning `err` itself?

Reply:
Couldn't find a protobuf type for error.
We do however return the error along with the response, we need to see what happens in the python sound when an error is returned along with the response - maybe it raises an exception?
@Tankilevitch @quaark do you have an idea?